### PR TITLE
Consul + Vault - TLS

### PR DESCRIPTION
**Pull request by danielehc**
Lab for TLS Encryption with Consul, Vault and consul-template.

### CODE REVIEW

File: consul-secure-tls-vault/index.json [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        ],
+        "terminals": [
+            {"name": "consul-template", "target": "host01"},
+            {"name": "operator", "target": "host01"}
+        ]
+    },
+    "backend": {
+        "port": 8080,
+        "imageid": "ubuntu"
+    }
```
Comment (yhyakuna) [inline consul-secure-tls-vault/index.json:91]:
```suggestion
    },
    "private": true
```

File: consul-secure-tls-vault/step2.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
+<!-- How would you feel about moving this before the Consul template step? Then in the Consul template step they can see the cert being rotated 
+
+We could move the configuration into a specific step and explain at the bottom that we are going to automate the config part using consul-template
+
+-->
+Now configure Consul using the `server.json`{{open}}
```
Comment (yhyakuna) [inline consul-secure-tls-vault/step2.md:6]:
```suggestion
Now configure Consul using the `server-tls.json`{{open}}
```

**Review by yhyakuna — APPROVED** [review]
Found a minor typo, but other than that, it looks good to me.

File: consul-secure-tls-vault/step2.md (hunk 2 above)
Comment (danielehc) [reply consul-secure-tls-vault/step2.md:6]: Very good catch, thanks.

File: consul-secure-tls-vault/finish.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
+## Review
+
+In this hands-on lab, you deployed a secure Consul datacenter using Vault to generate and manage certificates.
+
+The lab will guided you through the steps necessary to deploy Consul with TLS encryption enabled to secure access to the UI, API, CLI, services, and agents.
```
Comment (kaitlincarter-hc) [inline consul-secure-tls-vault/finish.md:5]:
```suggestion
The lab guided you through the steps necessary to deploy Consul with TLS encryption enabled to secure access to the UI, API, CLI, services, and agents.
```

File: consul-secure-tls-vault/finish.md [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+Specifically, you:
+
+- Started a Vault dev instance
+- Created a policy in Vault to allow certificate generation
+- Enabled the PKI engine in Vault
```
Comment (kaitlincarter-hc) [inline consul-secure-tls-vault/finish.md:11]:
```suggestion
- Enabled the PKI secrets engine in Vault
```

File: consul-secure-tls-vault/finish.md [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+- Initialized the CA and generate an intermediate certificate
+- Generated certificates for your Consul servers
+- Used consul-template to retrieve certificates at runtime
+- Performed a certificate rotation
+
+## Next Steps
+
+If you are already familiar with the basics of Consul, [Secure Consul with ACLs](https://learn.hashicorp.com/consul/security-networking/production-acls) provides a reference guide for the steps required to enable and use ACLs on your Consul datacenter.
```
Comment (kaitlincarter-hc) [inline consul-secure-tls-vault/finish.md:19]:
```suggestion
Now that you have tested TLS encryption in this interactive environment, use the (https://learn.hashicorp.com/tutorials/consul/tls-encryption-secure) tutorial to configure TLS encryption in your Consul datacenter. 
```

File: consul-secure-tls-vault/index.json [hunk 6]
```diff
@@ -0,0 +1,93 @@
+{
+    "pathwayTitle": "Generate mTLS Certificates for Consul using Vault",
```
Comment (kaitlincarter-hc) [inline consul-secure-tls-vault/index.json:2]:
```suggestion
    "pathwayTitle": "Generate Certificates with Vault's PKI Secrets Engine for Consul TLS Encryption",
```

File: consul-secure-tls-vault/index.json [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    "icon": "fa-consul",
+    "title": "Generate mTLS Certificates for Consul using Vault",
```
Comment (kaitlincarter-hc) [inline consul-secure-tls-vault/index.json:4]:
```suggestion
    "title": "Generate Certificates with Vault's PKI Secrets Engine for Consul TLS Encryption",
```

File: consul-secure-tls-vault/index.json [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    "description": "Deploy a Consul datacenter that uses Vault to generate and manage certificates to secure agent communication",
```
Comment (kaitlincarter-hc) [inline consul-secure-tls-vault/index.json:5]:
```suggestion
    "description": "Deploy a Consul datacenter that uses Vault's PKI secrets engine to generate and manage certificates to secure agent communication",
```

File: consul-secure-tls-vault/index.json [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    "details": {
+        "steps": [
+            {
+                "title": "Provision Lab Infrastructure",
```
Comment (kaitlincarter-hc) [inline consul-secure-tls-vault/index.json:9]:
```suggestion
                "title": "Lab Infrastructure Provisioning",
```

File: consul-secure-tls-vault/index.json [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+                "code": "provision.code.sh",
+                "verify": "provision.verify.sh"
+            },
+            {
+                "title": "Start Vault and enable the PKI secrets engine",
+                "code": "set-env.sh",
+                "text": "step0.md"
+            },
+            {
+                "title": "Configure Vault as CA",
```
Comment (kaitlincarter-hc) [inline consul-secure-tls-vault/index.json:21]:
```suggestion
                "title": "Configure Vault as Consul's CA",
```

File: consul-secure-tls-vault/intro.md [hunk 11]
```diff
@@ -0,0 +1,17 @@
+# Secure Consul Agent Communication with TLS
+
+In this hands-on lab, you will deploy a Consul datacenter that uses Vault to generate and manage certificates to secure agent communication.
+
+The lab will guide you through the steps necessary to deploy Consul with TLS encryption enabled to secure agent-to-agent communication including RPC requests and the consensus protocol.
```
Comment (kaitlincarter-hc) [inline consul-secure-tls-vault/intro.md:5]:
```suggestion
This lab will guide you through the steps necessary to deploy Consul with TLS encryption enabled to secure agent-to-agent communication including RPC requests and the consensus protocol.
```

File: consul-secure-tls-vault/intro.md [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+
+Specifically, you will:
+
+- Start a Vault dev instance
+- Create a policy in Vault to allow certificate generation
+- Enable the PKI secret engine in Vault
```
Comment (kaitlincarter-hc) [inline consul-secure-tls-vault/intro.md:11]:
```suggestion
- Enable the PKI secrets engine in Vault
```

File: consul-secure-tls-vault/step1.md [hunk 13]
```diff
@@ -0,0 +1,153 @@
+Consul requires that all clients and servers have key pairs that are generated by a single Certificate Authority. This can be a private CA, used only internally. The CA then signs keys for each of the agents.
```
Comment (kaitlincarter-hc) [inline consul-secure-tls-vault/step1.md:1]:
```suggestion
Consul requires that all servers and clients have key pairs that are generated by a single Certificate Authority. 
```

File: consul-secure-tls-vault/step1.md [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+
+In this lab you will use the PKI engine to generate the necessary CA and certificates.
```
Comment (kaitlincarter-hc) [inline consul-secure-tls-vault/step1.md:3]:
```suggestion
In this lab you will use the PKI secrets engine to generate the necessary CA and certificates.
```

File: consul-secure-tls-vault/step1.md [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+
+## Generate root CA
```
Comment (kaitlincarter-hc) [inline consul-secure-tls-vault/step1.md:5]:
```suggestion
## Generate the root CA
```

File: consul-secure-tls-vault/step2.md [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+}
+```
+
+To configure TLS encryption for Consul, three files are required:
+
+* `ca_file`   - CA (or intermediate) certificate to verify the identity of the other nodes.
+* `cert_file` - Consul agent public certificate
+* `key_file`  - Consul agent private key
+
+For the first Consul startup you will use the certificate generated earlier.
```
Comment (kaitlincarter-hc) [inline consul-secure-tls-vault/step2.md:31]:
```suggestion
For the first Consul startup, you will use the certificate generated earlier.
```

File: consul-secure-tls-vault/step2.md [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+
+You can manually create the files using the output of the command before now stored in `certs.txt` or you can use the following commands to extract the content from the file and place it into the right file and location.
```
Comment (kaitlincarter-hc) [inline consul-secure-tls-vault/step2.md:33]:
```suggestion
Use the following commands to extract the two certificates and private key from the `certs.txt` and place them into the right file and location.
```

File: consul-secure-tls-vault/step3.md [hunk 18]
```diff
@@ -0,0 +1,41 @@
+Once configuration is created and certificates and key are in place you can start Consul.
```
Comment (kaitlincarter-hc) [inline consul-secure-tls-vault/step3.md:1]:
```suggestion
Once the agent configuration is created and thee certificates are distributed, you can start Consul.
```

File: consul-secure-tls-vault/step3.md [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+
+`mkdir -p /opt/consul/data`{{execute T1}}
+
+`consul agent \
+     -config-file server.json \
+     -config-file server-tls.json \
+     -advertise '{{ GetInterfaceIP "ens3" }}' \
+     -data-dir=/opt/consul/data`{{execute T1}}
+
+In the output, if the configuration was successful,
```
Comment (kaitlincarter-hc) [inline consul-secure-tls-vault/step3.md:13]:
```suggestion
In the output
```

File: consul-secure-tls-vault/step3.md [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+you will get an indication in the output that TLS encryption is now enabled:
```
Comment (kaitlincarter-hc) [inline consul-secure-tls-vault/step3.md:14]:
```suggestion
The output will indicated that TLS encryption is now enabled:
```

File: consul-secure-tls-vault/step3.md [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+            Server: true (Bootstrap: true)
+       Client Addr: [127.0.0.1] (HTTP: 8500, HTTPS: 8501, gRPC: -1, DNS: 8600)
+      Cluster Addr: 172.17.0.32 (LAN: 8301, WAN: 8302)
+           Encrypt: Gossip: false, TLS-Outgoing: true, TLS-Incoming: true, Auto-Encrypt-TLS: true
+
+==> Log data will now stream in as it occurs:
+...
+```
+
+At this point in the lab you have a Consul server with TLS encryption configured.
```
Comment (kaitlincarter-hc) [inline consul-secure-tls-vault/step3.md:32]:
```suggestion
At this point, you have a Consul server with TLS encryption configured.
```

File: consul-secure-tls-vault/step4.md [hunk 22]
```diff
@@ -0,0 +1,95 @@
+You can use consul-template in your Consul datacenter to
+integrate with Vault's PKI Secrets Engine to generate
```
Comment (kaitlincarter-hc) [inline consul-secure-tls-vault/step4.md:2]:
```suggestion
integrate with Vault's PKI secrets engine to generate
```

File: consul-secure-tls-vault/step4.md [hunk 23]
```diff
@@ -0,0 +1,95 @@
...
+
+### Create and populate the templates directory
+
+This lab will demonstrate the TLS certificate automation
+for the server instances so that you can deploy a Consul
+datacenter that will generate and retrieve certificates
+from Vault and configure the servers automatically.
+
+You need to create templates that consul-template can use
+to render the actual certificates and keys on the nodes in
```
Comment (kaitlincarter-hc) [inline consul-secure-tls-vault/step4.md:13]:
```suggestion
to render the actual certificates and keys on the agent in
```

File: consul-secure-tls-vault/step4.md [hunk 24]
```diff
@@ -0,0 +1,95 @@
...
+your cluster. In this lab, you will place these templates
```
Comment (kaitlincarter-hc) [inline consul-secure-tls-vault/step4.md:14]:
```suggestion
your datacenter. In this lab, you will place these templates
```

File: consul-secure-tls-vault/step4.md [hunk 25]
```diff
@@ -0,0 +1,95 @@
...
+
+Example content:
+
+```
+{{ with secret "pki_int/issue/consul-dc1" "common_name=server.dc1.consul" "ttl=2m" "alt_names=localhost" "ip_sans=127.0.0.1"}}
+{{ .Data.certificate }}
+{{ end }}
+```
+
+The template will reach to the `pki_int/issue/consul-dc1` endpoint that Vault exposes to generate new certificates. It also mentions the common name and alternate names for the certificate.
```
Comment (kaitlincarter-hc) [inline consul-secure-tls-vault/step4.md:41]:
```suggestion
The template will use the `pki_int/issue/consul-dc1` endpoint that Vault exposes to generate new certificates. It also mentions the common name and alternate names for the certificate.
```

**Pull request merged**